src/Type/Php/CallUserFuncFunctionDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,31 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Type\CallableType;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\Type;
+
+class CallUserFuncFunctionDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'call_user_func';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		$callbackType = $scope->getType($functionCall->args[0]->value);
+
+		if ($callbackType instanceof CallableType) {
+			return $callbackType->getReturnType();
```

Comment:
This solves only a very narrow usecase when you have `callable $cb` as a parameter typehint. But callables can also be: arrays, strings, objects. This extension could be implemented as:

```php
$args = $functionCall->args;
// TODO check $args isn't empty - to prevent crash on `call_user_func()`
$firstArg = array_shift($args);
return $scope->getType(new FuncCall($firstArg->value, $args));
```

And all these cases would be taken care of thanks to this logic: https://github.com/phpstan/phpstan-src/blob/master/src/Analyser/MutatingScope.php#L1810-L1821

Reply:
This is nice, thanks.